app/controllers/devise/registrations_controller.rb

```diff
@@ -145,4 +141,30 @@ def account_update_params
   def translation_scope
     'devise.registrations'
   end
+
+  private
+
+  def set_flash_message_for_update(resource, prev_unconfirmed_email)
+    return unless is_flashing_format?
+    flash_key = if update_needs_confirmation?(resource, prev_unconfirmed_email)
+                  :update_needs_confirmation
+                elsif sign_in_after_change_password?
+                  :updated
+                else
+                  :updated_not_sign_in
+                end
+    set_flash_message :notice, flash_key
+  end
+
+  def sign_in_after_change_password
+    if sign_in_after_change_password?
+      bypass_sign_in resource, scope: resource_name
+    else
+      sign_out(resource)
+    end
+  end
+
+  def sign_in_after_change_password?
+    Devise.sign_in_after_change_password && account_update_params.include?(:password)
```

Comment:
Would `resource.password_changed?` work here? I ask because I'm worried we might return true here when the `password` parameter comes as an empty string, for example. The `#include` clause here will return true, but in fact, no `password` was provided. Makes sense?

Reply:
`sign_in_after_change_password?` is called only when [`resource_updated`]( https://github.com/plataformatec/devise/pull/4569/files#diff-f82eaf0d25964c41f76c557b038284e5R52), so we do not need to check whether blank or not?

Reply:
But then how would we know whether the password changed or not? I mean, if it's blank we assume that the user didn't want to change it, so in that case there would be no validation errors.